tools/clang/tools/hlsl2spirv/hlsl2spirv.cpp

```diff
@@ -0,0 +1,216 @@
+  void ReportResults();
+};
+
+void Hlsl2SpirvContext::ReportResults() {
+  bool needDisassembly =
+      !options.AssemblyCode.empty() || options.OutputObject.empty();
+
+  if (needDisassembly)
+    DisassembleResults();
+
+  if (!options.AssemblyCode.empty())
+    WriteBlobToFile(spirvAssemblyBlobPtr, StringRefUtf16(options.AssemblyCode));
+
+  if (!options.OutputObject.empty())
+    WriteBlobToFile(binaryBlobPtr, StringRefUtf16(options.OutputObject));
+  else
+    WriteUtf8ToConsole(spirvAssembly.c_str(), spirvAssembly.size(),
+                       STD_OUTPUT_HANDLE);
+}
+
+void Hlsl2SpirvContext::DisassembleResults() {
+  // First, convert the compiled blob into a vector of uint32_t
+  std::vector<uint32_t> words;
+  ConvertIDxcBlobToUint32(binaryBlobPtr, &words);
+
+  spvtools::SpirvTools spirvTools(SPV_ENV_UNIVERSAL_1_0);
+  spirvTools.SetMessageConsumer(
+      [](spv_message_level_t, const char *, const spv_position_t &,
+         const char *message) { fprintf(stdout, "%s\n", message); });
+  uint32_t options = SPV_BINARY_TO_TEXT_OPTION_FRIENDLY_NAMES;
+  spirvTools.Disassemble(words, &spirvAssembly, options);
+
+  // SPIR-V Tools returns the disassembly as a string. Convert to blob.
+  CComPtr<IDxcLibrary> library;
+  IFT(dllSupport.CreateInstance(CLSID_DxcLibrary, &library));
+  IFT(library->CreateBlobWithEncodingOnHeapCopy(
+      spirvAssembly.data(), spirvAssembly.size(), CP_UTF8,
+      (IDxcBlobEncoding **)&spirvAssemblyBlobPtr));
+}
+
+void Hlsl2SpirvContext::Compile() {
+  CComPtr<IDxcLibrary> pLibrary;
+  CComPtr<IDxcCompiler> pCompiler;
+  CComPtr<IDxcOperationResult> pResult;
+  CComPtr<IDxcBlobEncoding> pSource;
+  CComPtr<IDxcBlobEncoding> pErrorBuffer;
+  CComPtr<IDxcIncludeHandler> pIncludeHandler;
+  HRESULT resultStatus;
+
+  auto inputFileName = StringRefUtf16(options.InputFile);
+  auto entry = StringRefUtf16(options.EntryPoint);
+  auto profile = StringRefUtf16(options.TargetProfile);
+
+  std::vector<LPCWSTR> flags;
+  flags.push_back(L"-spirv");
+
+  IFT(dllSupport.CreateInstance(CLSID_DxcLibrary, &pLibrary));
+  IFT(pLibrary->CreateBlobFromFile(inputFileName, nullptr, &pSource));
+  IFT(pLibrary->CreateIncludeHandler(&pIncludeHandler));
+  IFT(dllSupport.CreateInstance(CLSID_DxcCompiler, &pCompiler));
+  IFT(pCompiler->Compile(pSource, inputFileName, entry, profile, flags.data(),
+                         flags.size(), nullptr, 0, pIncludeHandler, &pResult));
+  // Get compilation results.
+  IFT(pResult->GetStatus(&resultStatus));
+
+  // Get diagnostics string and print warnings and errors to stderr.
+  IFT(pResult->GetErrorBuffer(&pErrorBuffer));
+  const std::string diagnostics((char *)pErrorBuffer->GetBufferPointer(),
+                                pErrorBuffer->GetBufferSize());
+  fprintf(stderr, "%s\n", diagnostics.c_str());
+
+  if (SUCCEEDED(resultStatus)) {
+    IFT(pResult->GetResult(&binaryBlobPtr));
+  } else {
+    throw("Compilation of the input HLSL via dxc failed.");
```

Comment:
Maybe we should return the value from `pResult->GetStatus()` to `main()` and then also return there?